Source/DafnyLanguageServer/Workspace/DiagnosticPublisher.cs

```diff
@@ -26,11 +29,39 @@ public class DiagnosticPublisher : IDiagnosticPublisher {
     }
 
     private void PublishDocumentDiagnostics(DafnyDocument document) {
-      languageServer.TextDocument.PublishDiagnostics(new PublishDiagnosticsParams {
+      var diagnosticParameters = new PublishDiagnosticsParams {
         Uri = document.Uri,
         Version = document.Version,
         Diagnostics = GetDiagnostics(document).ToArray(),
-      });
+      };
+      languageServer.TextDocument.PublishDiagnostics(diagnosticParameters);
+    }
+
+    public void PublishVerificationDiagnostics(DafnyDocument document) {
+      if (document.LoadCanceled) {
+        // We leave the responsibility to shift the error locations to the LSP clients.
+        // Therefore, we do not republish the errors when the document (re-)load was canceled.
+        return;
+      }
+      // Document.GetDiagnostics() returns not only resolution errors, but previous verification errors
+      var currentDiagnostics =
+        document.Errors.GetDiagnostics(document.GetFilePath())
+          .Where(x => x.Severity == DiagnosticSeverity.Error).ToList();
+      var errors = currentDiagnostics
+        .Concat(document.OldVerificationDiagnostics)
+        .Where(x => x.Severity == DiagnosticSeverity.Error)
+        .ToArray();
+      var linesCount = Regex.Matches(document.Text.Text, "\r?\n").Count + 1;
+      var verificationStatusGutter = new VerificationStatusGutter(
+        document.Uri,
+        document.Version,
+        document.VerificationTree.Children.Select(child => child.GetCopyForNotification()).ToArray(),
+        errors,
```

Comment:
Why do you include errors here? Isn't publishing LSP diagnostics what `PublishDiagnosticsParams` is for?

Reply:
These errors are part of the constructor, not of the information sent to the client. I use these errors to render resolution errors on the gutter, if any.